### PR TITLE
make custom inline headers bootstrap configurable

### PR DESCRIPTION
**Pull request by wbpcode**
Signed-off-by: wbpcode <wbphub@live.com>

Commit Message: make custom inline headers bootstrap configurable
Additional Description:

Make custom inline headers bootstrap configurable. Check #17234 for more background info.

Risk Level: Mid.
Testing: Added.
Docs Changes: N/A.
Release Notes: Wait.

### CODE REVIEW

File: test/server/server_test.cc [hunk 1]
```diff
@@ -334,6 +334,47 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, ServerInstanceImplTest,
...
+}
+
+// Test whether the custom inline headers can be registered correctly.
+TEST_P(ServerInstanceImplTest, WithCustomInlineHeaders) {
+  static bool is_registered = false;
+
+  if (!is_registered) {
+    // The finalize() of Http::CustomInlineHeaderRegistry will be called after the first successful
+    // instantiation of InstanceImpl. If InstanceImpl is instantiated again and the inline header is
+    // registered again, the assertion will be triggered.
```
Comment (adisuissa) [inline test/server/server_test.cc:355]: Is there a test that covers this scenario? If not, can you please add one.

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 2]
```diff
@@ -578,3 +583,22 @@ message LayeredRuntime {
...
+    REQUEST_TRAILER = 1;
+    RESPONSE_HEADER = 2;
+    RESPONSE_TRAILER = 3;
+  }
+
+  // The name of the header that is expected to be set as the inline header.
+  string inline_header_name = 1 [(validate.rules).string = {min_len: 1}];
+
+  // The type of the header that is expected to be set as the inline header.
+  InlineHeaderType inline_header_type = 2;
```
Comment (adisuissa) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:603]: I think that the following PGV constraint `(validate.rules).enum.defined_only = true` can be added.

**Review by adisuissa — COMMENTED** [review]
Left an API comment and some drive-by code comments.

File: source/server/server.cc [hunk 3]
```diff
@@ -294,6 +295,46 @@ void loadBootstrap(absl::optional<uint32_t> bootstrap_version,
     throw EnvoyException(fmt::format("Unknown bootstrap version {}.", *bootstrap_version));
   }
 }
+
+bool canBeRegisteredAsInlineHeaderOrNot(const Http::LowerCaseString& header_name) {
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply source/server/server.cc:299]: 👌

File: source/server/server.cc [hunk 4]
```diff
@@ -294,6 +295,46 @@ void loadBootstrap(absl::optional<uint32_t> bootstrap_version,
...
+  if (header_name == Http::Headers::get().SetCookie) {
+    return false;
+  }
+  return true;
+}
+
+void registerCustomInlineHeadersFromBootstrap(
+    const envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+  for (const auto& inline_header : bootstrap.inline_headers()) {
+    Http::LowerCaseString lower_case_name(inline_header.inline_header_name());
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply source/server/server.cc:310]: 👌

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 2 above)
Comment (wbpcode) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:603]: 👌

File: test/server/server_test.cc (hunk 1 above)
Comment (wbpcode) [reply test/server/server_test.cc:355]: 👌

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Comment by wbpcode** [comment]
@alyssawilk Now the API is ready. When it is convenient for you, can you advance this work? 😄

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 5]
```diff
@@ -322,6 +322,11 @@ message Bootstrap {
   // field.
   // [#not-implemented-hide:]
   map<string, core.v3.TypedExtensionConfig> certificate_provider_instances = 25;
+
+  // Specifies a set of headers that need to be registered as inline header. This configuration
+  // allows users to customize the inline headers on-demand at Envoy startup without modifying
+  // Envoy's source code.
+  repeated CustomInlineHeader inline_headers = 32;
```
Comment (alyssawilk) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:329]: Worth calling out that inline headers will be elided?  If I recall correctly
 
foo: bar
foo: eep

will turn into

foo: bar, eep

File: source/server/server.cc [hunk 6]
```diff
@@ -292,6 +293,46 @@ void loadBootstrap(absl::optional<uint32_t> bootstrap_version,
     throw EnvoyException(fmt::format("Unknown bootstrap version {}.", *bootstrap_version));
   }
 }
+
+bool canBeRegisteredAsInlineHeader(const Http::LowerCaseString& header_name) {
```
Comment (alyssawilk) [inline source/server/server.cc:297]: call out in API?

File: api/envoy/config/bootstrap/v3/bootstrap.proto [hunk 7]
```diff
@@ -595,3 +600,22 @@ message LayeredRuntime {
...
+    REQUEST_TRAILER = 1;
+    RESPONSE_HEADER = 2;
+    RESPONSE_TRAILER = 3;
+  }
+
+  // The name of the header that is expected to be set as the inline header.
+  string inline_header_name = 1 [(validate.rules).string = {min_len: 1}];
+
+  // The type of the header that is expected to be set as the inline header.
+  InlineHeaderType inline_header_type = 2 [(validate.rules).enum = {defined_only: true}];
```
Comment (alyssawilk) [inline api/envoy/config/bootstrap/v3/bootstrap.proto:620]: vadd pgv for well_known_regex: HTTP_HEADER_NAME?

File: source/server/server.cc [hunk 8]
```diff
@@ -292,6 +293,46 @@ void loadBootstrap(absl::optional<uint32_t> bootstrap_version,
...
+void registerCustomInlineHeadersFromBootstrap(
+    const envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+  for (const auto& inline_header : bootstrap.inline_headers()) {
+    const Http::LowerCaseString lower_case_name(inline_header.inline_header_name());
+    if (!canBeRegisteredAsInlineHeader(lower_case_name)) {
+      throw EnvoyException(fmt::format("Header {} cannot be registered as an inline header.",
+                                       inline_header.inline_header_name()));
+    }
+    switch (inline_header.inline_header_type()) {
+    case envoy::config::bootstrap::v3::CustomInlineHeader::REQUEST_HEADER:
```
Comment (alyssawilk) [inline source/server/server.cc:314]: Can one add as a custom inline header a header which is already an inline header?  I think it's worth a unit test for this.

File: test/server/server_test.cc [hunk 9]
```diff
@@ -334,6 +334,60 @@ INSTANTIATE_TEST_SUITE_P(IpVersions, ServerInstanceImplTest,
...
+  static bool is_registered = false;
+
+  if (!is_registered) {
+    // Avoid repeated registration of custom inline headers in the current process after the
+    // finalize() of Http::CustomInlineHeaderRegistry has already been called.
+    EXPECT_NO_THROW(initialize("test/server/test_data/server/bootstrap_inline_headers.yaml"));
+    is_registered = true;
+  }
+
+  EXPECT_TRUE(
```
Comment (alyssawilk) [inline test/server/server_test.cc:373]: how about an integration test (protocol integration test?) that the headers / trailers end up elided?

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 5 above)
Comment (alyssawilk) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:329]: please call out the new field in docs/root/version_history/current.rst

**Review by alyssawilk — COMMENTED** [review]
Looks great! this will be a really useful feature going forward :-)

File: source/server/server.cc (hunk 6 above)
Comment (wbpcode) [reply source/server/server.cc:297]: get it.

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 5 above)
Comment (wbpcode) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:329]: get it.

File: api/envoy/config/bootstrap/v3/bootstrap.proto (hunk 7 above)
Comment (wbpcode) [reply api/envoy/config/bootstrap/v3/bootstrap.proto:620]: get it.

File: source/server/server.cc (hunk 8 above)
Comment (wbpcode) [reply source/server/server.cc:314]: > Can one add as a custom inline header a header which is already an inline header?
Yes.

> I think it's worth a unit test for this.

I think this scenario has been verified in the test of `CustomInlineHeaderRegistry::registerInlineHeader` itself. We can register the header with the same name as the inline header repeatedly without any negative effects. 

https://github.com/envoyproxy/envoy/blob/d483098022e72de39a23dc5f9ed65a044ae86b3a/test/common/http/header_map_impl_test.cc#L365-L368

https://github.com/envoyproxy/envoy/blob/d483098022e72de39a23dc5f9ed65a044ae86b3a/test/common/http/header_map_impl_test.cc#L389-l399

File: test/server/server_test.cc (hunk 9 above)
Comment (wbpcode) [reply test/server/server_test.cc:373]: get it. I think we can verify this problem with a few simple lines of tests in existing test cases.

**Review by alyssawilk — APPROVED** [review]
Looks great!  Throwing over to @snowp for second pass

**Comment by alyssawilk** [comment]
Also @adisuissa can you do another API stamp?

**Review by adisuissa — COMMENTED** [review]
/lgtm api

File: test/server/server_test.cc (hunk 9 above)
Comment (snowp) [reply test/server/server_test.cc:373]: Where did we land on this? Don't see an integration test

**Review by snowp — COMMENTED** [review]
This LGTM, just one question re adding an integration test

**Comment by wbpcode** [comment]
@snowp Thanks.

@alyssawilk's suggestion is to verify that a header or trailer registered as an inline header can be folded correctly. In practice, integration tests are not necessarily required to verify this problem. 
Instead of adding a whole new integration test to verify it, I did the same by adding some extra code to the test case.

https://github.com/wbpcode/envoy/blob/446214acbbce602b8b9a84ad1b0e342c65e47535/test/server/server_test.cc#L390-L427

**Comment by snowp** [comment]
I think normally we encourage adding an integration test even if it's not strictly necessary since they tend to catch unexpected issues (or provide more reliable regression tests). That said it seems like we're okay skipping it in this case, so let's get this merged

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**